core/src/main/scala/chisel3/probe/ProbeValue.scala

```diff
@@ -2,24 +2,46 @@
 
 package chisel3.probe
 
-import chisel3.{Data, SourceInfoDoc}
+import chisel3.{Data, Output, SourceInfoDoc, Wire}
 import chisel3.internal.{Builder, OpBinding}
-import chisel3.internal.firrtl.{ProbeExpr, RWProbeExpr}
+import chisel3.internal.Builder.pushCommand
+import chisel3.internal.firrtl.{ProbeDefine, ProbeExpr, RWProbeExpr}
 import chisel3.experimental.{requireIsHardware, SourceInfo}
+import chisel3.reflect.DataMirror.collectLeafMembers
 
 import scala.language.experimental.macros
+import chisel3.Aggregate
 
 private[chisel3] sealed trait ProbeValueBase {
   protected def apply[T <: Data](source: T, writable: Boolean)(implicit sourceInfo: SourceInfo): T = {
     requireIsHardware(source)
     // construct probe to return with cloned info
-    val clone = if (writable) RWProbe(source.cloneType) else Probe(source.cloneType)
+    val clone = if (writable) RWProbe(source.cloneTypeFull) else Probe(source.cloneTypeFull)
     clone.bind(OpBinding(Builder.forcedUserModule, Builder.currentWhen))
-    if (writable) {
-      clone.setRef(RWProbeExpr(source.ref))
+    // create reference for clone
+    val cloneRef = if (source.isInstanceOf[Aggregate]) {
+      // intermediate probe to hook up to aggregate elements
+      val intermediate = Wire(Output(if (writable) RWProbe(source.cloneTypeFull) else Probe(source.cloneTypeFull)))
+      collectLeafMembers(intermediate).zip(collectLeafMembers(source)).foreach {
```

Comment:
similar to my comments above but obviously you don't need to check correctness. That being said, why can't this code use the same logic as in `secretProbeDefines` (which still feels to me like it should just be the normal connections logic. It's the same principle even if this code won't be able to exercise all code paths.

Also, it would then just handle the leaf case with the same logic I think.

Reply:
yes again i think the `DataMirror.collectMembersOverMatches` is the suggested API here